src/compute/src/render/top_k.rs

```diff
@@ -317,6 +332,143 @@ where
+                                let mut session = output.session(&time);
+                                for (record_time, topk) in aggs {
+                                    session.give_iterator(topk.into_iter().map(
+                                        |(monoid, diff)| (monoid.row, record_time.clone(), diff),
+                                    ))
+                                }
+                            }
+                        });
+                    },
+                )
+                .as_collection()
+        }
+    }
+}
+
+/// Types for in-place intra-ts aggregation of monotonic streams.
+pub mod topk_agg {
+    use smallvec::SmallVec;
+
+    pub struct TopKBatch<T> {
+        updates: SmallVec<[(T, i64); 16]>,
+        clean: usize,
+        limit: i64,
+    }
+
+    impl<T: Ord> TopKBatch<T> {
+        pub fn new(limit: i64) -> Self {
+            Self {
+                updates: SmallVec::new(),
+                clean: 0,
+                limit,
+            }
+        }
+
+        /// Adds a new update, for `item` with `value`.
+        ///
+        /// This could be optimized to perform compaction when the number of "dirty" elements exceeds
+        /// half the length of the list, which would keep the total footprint within reasonable bounds
+        /// even under an arbitrary number of updates. This has a cost, and it isn't clear whether it
+        /// is worth paying without some experimentation.
+        #[inline]
+        pub fn update(&mut self, item: T, value: i64) {
+            self.updates.push((item, value));
+            self.maintain_bounds();
+        }
+
+        /// Compact the internal representation.
+        ///
+        /// This method sort `self.updates` and consolidates elements with equal item, discarding
+        /// any whose accumulation is zero. It is optimized to only do this if the number of dirty
+        /// elements is non-zero.
+        #[inline]
+        pub fn compact(&mut self) {
+            if self.clean < self.updates.len() && self.updates.len() > 1 {
+                self.updates.sort_by(|x, y| x.0.cmp(&y.0));
+                for i in 0..self.updates.len() - 1 {
+                    if self.updates[i].0 == self.updates[i + 1].0 {
+                        self.updates[i + 1].1 += self.updates[i].1;
+                        self.updates[i].1 = 0;
+                    }
+                }
+                let mut limit = self.limit;
+                self.updates.retain(|x| {
+                    if limit > 0 {
+                        limit -= x.1;
+                        true
+                    } else {
+                        false
+                    }
+                });
+                // Adjust the diff of the last record that was retained so that we have exactly K
+                // records
+                if let Some(item) = self.updates.last_mut() {
+                    item.1 -= -limit;
+                }
```

Comment:
I don't really follow this. It looks like up above `limit` could go negative if records have multiplicity greater than one, in which case this will increment the count? I assume that isn't what you want, but instead to subtract some trailing counts?

Reply:
This is confusing, maybe I should add more comments. I am subtracting the negated limit, so if `limit` goes negative I'm subtracting a positive number, which is subtracting the trailing counts. I had it as `item.1 += limit` originally but that seemed equally confusing to me.